### DISCUSSION TITLE
Recreating Steering limits setup in Blocks

### DISCUSSION THREAD

**Discussion by rbeckett75** [pybricks/support#1339] [Q&A] [answered]
I'm building a mechanism that needs to rotate a turntable, randomly between, say, 300' and 60'.   Since I don't know what angle it will be at, when starting, I figured I'd use a steering-style mechanism (I have limit-stops in the build), to work out the center.

I looked a the Powered up RC Car Tutorial, and saw this:

```
## Find the steering endpoint on the left and right.
## The middle is in between.
left_end = steer.run_until_stalled(-200, then=Stop.HOLD)
right_end = steer.run_until_stalled(200, then=Stop.HOLD)

## We are now at the right. Reset this angle to be half the difference.
## That puts zero in the middle.
steer.reset_angle((right_end - left_end) / 2)
steer.run_target(speed=200, target_angle=0, wait=False)
```

But after a bit of playing around in Blocks - can't see an easy way to recreate this?

thanks in advance!

**Comment by laurensvalk** [comment]
There's a few ways to do it:

1. You could do it almost line by line as above: use run to endstop, save the angle, run to the other end, and save that angle too. Then reset as above.

![image](https://github.com/pybricks/support/assets/12326241/d5c726e4-b885-48f2-b83a-152c8244c3cc)

![image](https://github.com/pybricks/support/assets/12326241/c79f6ede-ce52-4a1c-9f57-d793c3b47e66)

![image](https://github.com/pybricks/support/assets/12326241/530d47cc-5a90-46c2-a14c-cb06c2b2ac08)

Putting that together, you get:

![image](https://github.com/pybricks/support/assets/12326241/1a6a70dc-18a7-4ceb-9a0b-12445979df8e)


2. Notice that unlike in your code, the block version has no return value for `run_until_stalled`, so there is an extra line to read the angle. But you could create a function to mimic the run until stalled behavior with a return value. This way you don't need variables:

![image](https://github.com/pybricks/support/assets/12326241/93a2b361-32ab-46d5-9a84-4ac18f867a47)

3. You could wait until we release it as a dedicated block :smile:  This will essentially work as a mix between 1 and 2. It will add such a utility function and then call it. I've had this as one of the options for the run block. But since blocks were all new, I didn't want to add to many new/different functions at the same time (to keep things simple). It's definitely a nice utility, though, so I think it makes sense to add it down the line.

**Reply by rbeckett75** [reply]
Thanks for that!

I figured it might be something you'd be adding later
(Also, I couldn't find a way to to GEARS paramater for the motor block)


That said.. this morning, before work.. I tried both options.. and neither seemed to return the motor to centre after doing the calibration end-stops.

I tried it with just a bare motor, and a liftarm.. it goes to both ends... and then just runs back, and jams against the stop.
If I quickly removed the technic brick acting as an end-stop, after both calibration-stops, it seems to run back past it, and end up at a strange angle..

I'll recheck my code tonight.. but don't think I had anything obviously different

**Reply by laurensvalk** [reply, marked as answer]
If in doubt, compare to the Python code that was working. Carefully note left/right +/- conventions as well.

Other than doing the `run_until_stalled` and `angle` measurement in two steps, I think the first example is the same as yours.

**Reply by rbeckett75** [reply]
Yep.. must have made a mistake transcribing it... all working now... :)

now onto my next question!  (will start a new thread, as it's a different topic!)
